substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -95,16 +99,319 @@ private PosixJavaNetSubstitutions() {
+            fd = PosixUtils.getFD(fdObj);
+            //   203          fd = (*env)->GetIntField(env, fdObj, IO_fd_fdID);
+        }
+
+        //   206      if (IS_NULL(iaObj)) {
+        if (iaObj == null) {
+            //   207          JNU_ThrowNullPointerException(env, "iaObj is null.");
+            //   208          return;
+            throw new NullPointerException("iaObj is null.");
+        }
+
+        //   211      /* bind */
+        //   212      if (NET_InetAddressToSockaddr(env, iaObj, localport, (struct sockaddr *)&him, &len, JNI_TRUE) != 0) {
+        len_Pointer.write(0);
+        if (JavaNetNetUtilMD.NET_InetAddressToSockaddr(iaObj, localport, him, len_Pointer, Util_jni.JNI_TRUE()) != 0) {
+            //   213        return;
+            return;
+        }
+
+        //   215      setDefaultScopeID(env, (struct sockaddr *)&him);
+        JavaNetNetUtilMD.setDefaultScopeID(him);
+
+        //   217      if (NET_Bind(fd, (struct sockaddr *)&him, len) < 0)  {
+        if (JavaNetNetUtilMD.NET_Bind(fd, him, len_Pointer.read()) < 0) {
+            //   218          if (errno == EADDRINUSE || errno == EADDRNOTAVAIL ||
+            //   219              errno == EPERM || errno == EACCES) {
+            if (Errno.errno() == Errno.EADDRINUSE() || Errno.errno() == Errno.EADDRNOTAVAIL() || //
+                    Errno.errno() == Errno.EPERM() || Errno.errno() == Errno.EACCES()) {
+                //   220              NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "BindException",
+                //   221                              "Bind failed");
+                throw new BindException(PosixUtils.lastErrorString("Bind failed"));
+            } else {
+                //   223              NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+                //   224                              "Bind failed");
+                throw new SocketException(PosixUtils.lastErrorString("Bind failed"));
+            }
+        }
+
+        //   229      /* initialize the local port */
+        //   230      if (localport == 0) {
+        if (localport == 0) {
+            //   231          /* Now that we're a connected socket, let's extract the port number
+            //   232           * that the system chose for us and store it in the Socket object.
+            //   233           */
+            //   234          if (JVM_GetSockName(fd, (struct sockaddr *)&him, &len) == -1) {
+            if (VmPrimsJVM.JVM_GetSockName(fd, him, len_Pointer) == -1) {
+                //   235              NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
+                //   236                              "Error getting socket name");
+                try {
+                    JavaNetNetUtilMD.NET_ThrowNew(Errno.errno(), "Error getting socket name");
+                    //   237              return;
+                    return;
+                } finally {
+                    return;
+                }
```

Comment:
You do not need the `try .... finally`.
You could add a comment that says that the C line 237 is unreachable in the Java code.

Reply:
But the original C code is, as you quoted,
```
//   235              NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "SocketException",
//   236                              "Error getting socket name");
```
which can only throw a `SocketException`.  The translation should be
```
throw new SocketException(PosixUtils.lastErrorString("Error getting socket name"));
```
That matches the declaration of the Java method `PlainDatagramSocketImpl.bind0`, which says it only throws `SocketException`.

Reply:
Yah, I finally realized that. Some days I can be a tad slow on the uptake.